wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -301,11 +313,21 @@ public boolean hasRuntime(String workspaceId) {
     }
 
     @PostConstruct
+    private void init() {
+        recover();
+        subscribeCleanupOnAbnormalRuntimeStopEvent();
+    }
+
     private void recover() {
         for (RuntimeInfrastructure infra : infraByRecipe.values()) {
             try {
                 for (RuntimeIdentity id : infra.getIdentities()) {
-                    runtimes.put(id.getWorkspaceId(), validate(infra.getRuntime(id)));
+                    // TODO how to identify correct state of runtime
+                    if (runtimes.putIfAbsent(id.getWorkspaceId(),
+                                             new RuntimeState(validate(infra.getRuntime(id)), RUNNING)) != null) {
+                        // should not happen, violation of SPI contract
+                        LOG.error("More than 1 runtime of workspace with id '' found. Skipping all duplicates.");
```

Comment:
Missing _id_ parameter. I think more information should be logged, for example _RuntineIdentity_ of the duplicate